tensorflow/core/kernels/decode_bmp_op.cc

```diff
@@ -0,0 +1,125 @@
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace tensorflow {
+
+// Decode the contents of a BMP file
+class DecodeBmpOp : public OpKernel {
+ public:
+  explicit DecodeBmpOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& contents = context->input(0);
+    OP_REQUIRES(context, TensorShapeUtils::IsScalar(contents.shape()),
+                errors::InvalidArgument("contents must be scalar, got shape ",
+                                        contents.shape().DebugString()));
+
+    // Start decoding image to get shape details
+    const StringPiece input = contents.scalar<string>()();
+
+    // Decode image, allocating tensor once the image size is known
+    Tensor* output = nullptr;
+    OP_REQUIRES(
+        context,
+        Decode(input.data(), input.size(),
+                    [=, &output](int width, int height,
+                                 int channels) -> uint8* {
+                      Status status(context->allocate_output(
+                          0, TensorShape({height, width, channels}),
+                          &output));
+                      if (!status.ok()) {
+                        VLOG(1) << status;
+                        context->SetStatus(status);
+                        return nullptr;
+                      }
+                      return output->flat<uint8>().data();
+                    }),
+        errors::InvalidArgument("Invalid BMP data, size ", input.size()));
+  }
+
+  uint8* Decode(const void* srcdata, int datasize,
+              std::function<uint8*(int, int, int)> allocate_output);
+};
+REGISTER_KERNEL_BUILDER(Name("DecodeBmp").Device(DEVICE_CPU), DecodeBmpOp);
+
+uint8* DecodeBmpOp::Decode(const void* srcdata, int datasize,
+              std::function<uint8*(int, int, int)> allocate_output) {
+
+  bool top_down = false;
+  uint8* const img_bytes = (uint8 *) srcdata;
+  const int header_size = *(reinterpret_cast<int*>(img_bytes + 10));
+  const int width = *(reinterpret_cast<int*>(img_bytes + 18));
+  const int height = *(reinterpret_cast<int*>(img_bytes + 22));
+
+  // if height is negative, data layout is top down
+  // otherwise, it's bottom up
+  if (height < 0)
+    top_down = true;
+  uint8* dstdata = allocate_output(width, abs(height), 3);
+  uint8* const bmp_pixels = &img_bytes[header_size];
+  int row_size = (8 * width * 3 + 31) / 32 * 4;
+
+#ifndef NDEBUG
+  const int data_size = *(reinterpret_cast<int*>(img_bytes + 34));
+  LOG(INFO) << "header size = " << header_size;
+  LOG(INFO) << "width = " << width;
+  LOG(INFO) << "height = " << height;
+  LOG(INFO) << "data size = " << data_size;
+  LOG(INFO) << "row size = " << row_size;
+  LOG(INFO) << "top down = " << top_down;
+#endif
+
+  for (int i = 0; i < abs(height); i++) {
```

Comment:
I am not familiar at all with bmp, so I don't know :(.

Since BMP is a windows format, I would assume Microsoft's decoder would probably be a start ;) https://msdn.microsoft.com/en-us/library/system.windows.media.imaging.bmpbitmapdecoder(v=vs.110).aspx

My general feeling is that this function could be optimized in two related ways:

1) Using multiple threads, mapping each thread to a unique portion of the input.

2) Within each thread, handing it enough work such that it could maybe share some of the indexing computation for many output pixels.

It *might* make sense for you to play around with this on your own first, and to do some benchmark comparisons to other libraries.  At a point when you think this implementation would be somewhat competitive with other libraries, it might be a good candidate for inclusion in core TensorFlow.  How does that sound?